MSC3754: Removing profile information (#3754)
matrix-org/matrix-spec-proposals: proposals/3754-removing-profile-info.md

# MSC3754: Removing Profile Information

In the Client-Server API specification, there is currently no clear way to remove the profile information, or to know
when it has been removed.

## Proposal

On top of the current profile endpoints:

```http
GET /_matrix/client/v3/profile/{userId}/avatar_url
GET /_matrix/client/v3/profile/{userId}/displayname
PUT /_matrix/client/v3/profile/{userId}/avatar_url
PUT /_matrix/client/v3/profile/{userId}/displayname
```

This proposal adds two new endpoints for deleting the data:

```http
DELETE /_matrix/client/v1/profile/{userId}/avatar_url

> DELETE /_matrix/client/v1/profile/{userId}/displayname

Reviewer: Shouldn't these be `v3` in accordance with the existing endpoints?

Author: I'm not sure actually. From what I understand endpoint versioning should be separate, and changing the HTTP method makes a new endpoint.

Reviewer: (Maybe someone from SCT could help resolve this then)

Reviewer 2: Debatable. I'd vote for consistency with the existing endpoints in this case

Reviewer 3: new endpoints are meant to be v1, but that's not a rule which must be followed. Sensible choices > rules here, so v3 to match

```

To identify the change, the `m.room.member` and `m.presence` events sent following the call to one of these two
endpoints MUST include the changed key with a `null` value. Subsequent `m.room.member` events SHOULD omit the deleted

> key altogether.

Reviewer 4: Why must the `m.room.member` eventsinclude `null` instead of omitting the values? The current spec for [`m.room.member`](https://spec.matrix.org/v1.10/client-server-api/#mroommember) and [`m.presence`](https://spec.matrix.org/v1.10/client-server-api/#mpresence) for these fields is pretty clear:

>  The {avatar URL|display name} for this user, if any.

My reading of this is that the new values fully replace any old values and omitted values should be cleared.

Note that I think https://spec.matrix.org/v1.10/client-server-api/#events-on-change-of-profile-information is the useful part of the spec we're talking about? Cross-linking in the MSC is very much appreciated.

Author: This was changed and I added the link.

To reflect those changes, omitting the `avatar_url` field in the body of the request to `PUT […]/avatar_url` or
the `displayname` field in the body of the request to `PUT […]/displayname` is now deprecated. They can be marked as
required in a future MSC.

## Potential issues

Marking the fields in the body of the `PUT` endpoints requests as required would break compatibility for existing
clients so they are only deprecated for now. This means that the servers must accept two different ways of removing
profile data for the time being.

## Alternatives

Even if it is not documented, it is currently possible to remove profile account data. The general consensus is to send
`{ avatar_url: "" }` for `avatar_url` or `{}` for `displayname`. This means that a clarification in the spec would
either be inconsistent between the `PUT` endpoints or it whould require a change in the current implementations.

> The `m.room.member` event omits the changed key in `content`. This means that all the keys that are not unset should be
> present every time a part of the profile is changed. This is limiting for custom keys as they might be omitted by
> clients or servers that don't recognize them.

Reviewer 4: I'm not really following this, maybe fewer double negatives helps?

```suggestion
The `m.room.member` event omits the changed key in `content`. This means that all the keys that ar set should be
present every time a part of the profile is changed. This is limiting for custom keys as they might be omitted by
clients or servers that don't recognize them.
```

Author: This paragraph was removed.
